<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simulating GWAS summary data • simulateGP</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Simulating GWAS summary data">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">simulateGP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/fast_linreg.html">Performing fast linear regression</a>
    </li>
    <li>
      <a href="../articles/gwas_summary_data.html">Simulating GWAS summary data</a>
    </li>
    <li>
      <a href="../articles/individual_gwas.html">Simulating individual level GWAS data</a>
    </li>
    <li>
      <a href="../articles/risk.html">Translating between observed and liability models</a>
    </li>
    <li>
      <a href="../articles/simulate_individual_2smr.html">Simulate two-sample MR from individual level data</a>
    </li>
    <li>
      <a href="../articles/weak_instruments.html">Simulating summary data for weak instruments</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Simulating GWAS summary data</h1>
            
      
      
      <div class="hidden name"><code>gwas_summary_data.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(simulateGP)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="co">#&gt; Loading required package: tidyverse</span></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="co">#&gt; ── Attaching packages ────────────────────────────────── tidyverse 1.2.1 ──</span></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="co">#&gt; ✔ ggplot2 3.2.1     ✔ purrr   0.3.3</span></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="co">#&gt; ✔ tibble  2.1.3     ✔ dplyr   0.8.3</span></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="co">#&gt; ✔ tidyr   1.0.0     ✔ stringr 1.4.0</span></span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="co">#&gt; ✔ readr   1.3.1     ✔ forcats 0.4.0</span></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="co">#&gt; ── Conflicts ───────────────────────────────────── tidyverse_conflicts() ──</span></span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="co">#&gt; ✖ dplyr::filter() masks stats::filter()</span></span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="co">#&gt; ✖ dplyr::lag()    masks stats::lag()</span></span></code></pre></div>
<p>GWAS summary data can be generated in two ways</p>
<ol style="list-style-type: decimal">
<li>Given a set of parameters, generate the genotypes for a set of individuals and the phenotypes that arise from those genotypes. Then obtain linear regression estimates from these simulated individual level data</li>
<li>Given a set of parameters, simulate the summary level data directly</li>
</ol>
<p>here we look at (1).</p>
<p>Given the following inputs:</p>
<ol style="list-style-type: decimal">
<li>True causal effects of a set of SNPs</li>
<li>Minor allele frequency of those SNPs</li>
<li>Heritability explained by those SNPs</li>
<li>Sample size in which they are estimated</li>
</ol>
<p>We would like to obtain:</p>
<ol style="list-style-type: decimal">
<li>Standard error of SNP effect on trait</li>
<li>Sampled estimate of SNP effect on trait</li>
<li>P-value of SNP effect estimate for trait</li>
</ol>
<p>Strategy is to begin by obtaining the expected standard error for the true causal effect for each SNP, and then sampling an estimated SNP effect based on the expected standard error.</p>
<div id="expected-standard-error" class="section level2">
<h2 class="hasAnchor">
<a href="#expected-standard-error" class="anchor"></a>Expected standard error</h2>
<p>For a linear model</p>
<p><span class="math display">\[
y_i = \beta_0 + \beta_1 x_i + \epsilon_i
\]</span></p>
<p>The coefficient estimates are</p>
<p><span class="math display">\[
\hat{\beta}_1 = cov(x,y) / var(x)
\]</span></p>
<p>and</p>
<p><span class="math display">\[
\hat{\beta}_0 = \bar{y} - \hat{\beta}_1 \bar{x}
\]</span></p>
<p>The standard error of a regression coefficient is:</p>
<p><span class="math display">\[
s_{\hat{\beta}_1} = \sqrt{\frac{\sum{\epsilon_i^2}}{(n-2) \sum{(x_i - \bar{x})^2}}}
\]</span></p>
<p>This can be broken down. The denominator is the sum of squares of X:</p>
<p><span class="math display">\[
\begin{aligned}
(x_i - \bar{x})^2 &amp;= (n-1) Var(x) \\
&amp;\approx (n-1) 2p(1-p)
\end{aligned}
\]</span></p>
<p>where <span class="math inline">\(p\)</span> is the allele frequency of the SNP and <span class="math inline">\(n\)</span> is the sample size. The numerator is the mean squared error, which relates to the variance unexplained in Y:</p>
<p><span class="math display">\[
\begin{aligned}
\sum \epsilon_i^2 &amp;= \sum(y_i - \hat{y}_i)^2 / (n-1) \\
&amp;= Var(y) - r^2 Var(y) \\
&amp;\approx Var(y) - \hat{\beta}_1^2 Var(x) \\
&amp;\approx Var(y) - 2 p(1-p) \hat{\beta}_1^2
\end{aligned}
\]</span></p>
<p>Quick check:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>n &lt;-<span class="st"> </span><span class="dv">100</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>y &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(n)</span>
<span id="cb2-3"><a href="#cb2-3"></a>x &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(n)</span>
<span id="cb2-4"><a href="#cb2-4"></a></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="kw"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(y <span class="op">~</span><span class="st"> </span>x))<span class="op">$</span>coefficients[<span class="dv">2</span>,<span class="dv">2</span>]</span>
<span id="cb2-6"><a href="#cb2-6"></a><span class="co">#&gt; [1] 0.1120962</span></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="kw"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span>((<span class="kw"><a href="https://rdrr.io/r/stats/cor.html">var</a></span>(y) <span class="op">-</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(x,y)<span class="op">^</span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/cor.html">var</a></span>(y)) <span class="op">/</span><span class="st"> </span>(<span class="kw"><a href="https://rdrr.io/r/stats/cor.html">var</a></span>(x) <span class="op">*</span><span class="st"> </span>(n<span class="dv">-2</span>)))</span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="co">#&gt; [1] 0.1120962</span></span></code></pre></div>
</div>
<div id="sampling-the-effect-estimate" class="section level2">
<h2 class="hasAnchor">
<a href="#sampling-the-effect-estimate" class="anchor"></a>Sampling the effect estimate</h2>
<p>Given a standard error <span class="math inline">\(s_{\hat{\beta}_1}\)</span>, we can obtain an estimated value for <span class="math inline">\(\beta_1\)</span> using</p>
<p><span class="math display">\[
\hat{\beta}_1 \sim N(\beta_1, s_{\hat{\beta}_1})
\]</span></p>
<p>Question: Do I need to update the expected standard error based on the sampled value of <span class="math inline">\(\hat{\beta}_1\)</span>?</p>
</div>
<div id="sampling-a-set-of-effect-estimates" class="section level2">
<h2 class="hasAnchor">
<a href="#sampling-a-set-of-effect-estimates" class="anchor"></a>Sampling a set of effect estimates</h2>
<p>Suppose we need a set of SNPs to explain some proportion of the variance in a trait.</p>
<p>When specifying the SNP effects there is a simple model that can relate to a SNP to its contribution to heritability</p>
<p><span class="math display">\[
h^2_j = \frac{2p_j(1-p_j)\beta^2_j}{Var(y)}
\]</span></p>
<p>Also</p>
<p><span class="math display">\[
\begin{aligned}
h_j^2 &amp;= \frac{Cov(x_j, y)^2}{Var(x_j)Var(y)} \\
&amp;= \frac{\beta_j^2}{Var(y)}
\end{aligned}
\]</span></p>
<p>and</p>
<p><span class="math display">\[
h^2 = \sum^M_j h_j^2
\]</span></p>
<p>If each SNP has MAF of <span class="math inline">\(p_j\)</span> and effect of <span class="math inline">\(\beta_j\)</span>, then</p>
<p><span class="math display">\[
V_G = \sum^M_j 2p_j(1-p_j)\beta^2_j
\]</span></p>
<p><span class="math display">\[
V_E = \frac{V_G(1 - h^2)}{h^2}
\]</span> and</p>
<p><span class="math display">\[
Var(y) = V_G + V_E
\]</span></p>
<p>Once SNP effects have been sampled then they can be scaled to represent a phenotype with variance of 1:</p>
<p><span class="math display">\[
\beta_j^* = \frac{\beta_j}{\sqrt{Var(y)}}
\]</span></p>
<p>Relating SNP effects to allele frequencies can be done by specifying a selection model. Generalise the relationship between the SNP effects, allele frequencies and selection coefficient of the trait:</p>
<p><span class="math display">\[
\beta_j \sim N(0, [2p_j (1-p_j)]^S \sigma^2_\beta)
\]</span></p>
<p>Where S is the parameter describing the selection acting on the trait. Here <span class="math inline">\(\sigma^2_\beta = V_G / M\)</span> if all SNPs were scaled to have variance of 1 and mean of 0.</p>
<p><span class="math display">\[
\begin{aligned}
\sigma^2_\beta &amp;= \sum^M_j (\beta_j-\bar{\beta})^2 \\
&amp;= \sum^M_j \beta_j^2
\end{aligned}
\]</span></p>
<p>when the distribution of SNP effects is centred. This can be adapted to the BayesS model, in which some proportion of SNPs have no effect:</p>
<p><span class="math display">\[
\beta_j \sim N(0, [2p_j (1-p_j)]^S \sigma^2_\beta)\pi + \vartheta (1 - \pi)
\]</span></p>
<p>I think ultimately the presence of <span class="math inline">\(\sigma^2_\beta\)</span> in this model does not actually mean anything, it scales the SNP effects with respect to the variance of the phenotype. Use the following function to generate SNP effects for a set of SNPs according to specified MAF, heritability and selection model:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>generate_gwas_params</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="co">#&gt; function (maf, h2, S = 0) </span></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="co">#&gt; {</span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co">#&gt;     nsnp &lt;- length(maf)</span></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co">#&gt;     if (h2 == 0) {</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co">#&gt;         return(dplyr::tibble(beta = 0, maf = maf))</span></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="co">#&gt;     }</span></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="co">#&gt;     beta &lt;- rnorm(nsnp, mean = 0, sd = sqrt((maf * 2 * (1 - maf))^S))</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="co">#&gt;     vg &lt;- sum(maf * 2 * (1 - maf) * beta^2)</span></span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="co">#&gt;     ve &lt;- (vg - h2 * vg)/h2</span></span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="co">#&gt;     vy &lt;- vg + ve</span></span>
<span id="cb3-12"><a href="#cb3-12"></a><span class="co">#&gt;     beta &lt;- beta/sqrt(vy)</span></span>
<span id="cb3-13"><a href="#cb3-13"></a><span class="co">#&gt;     return(dplyr::tibble(beta = beta, maf = maf))</span></span>
<span id="cb3-14"><a href="#cb3-14"></a><span class="co">#&gt; }</span></span>
<span id="cb3-15"><a href="#cb3-15"></a><span class="co">#&gt; &lt;bytecode: 0x7fc456ac4348&gt;</span></span>
<span id="cb3-16"><a href="#cb3-16"></a><span class="co">#&gt; &lt;environment: namespace:simulateGP&gt;</span></span></code></pre></div>
<p>Negative selection:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>g &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span>(<span class="dt">maf=</span><span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="dv">1000</span>, <span class="fl">0.01</span>, <span class="fl">0.5</span>), <span class="dt">h2=</span><span class="fl">0.4</span>, <span class="dt">S=</span><span class="op">-</span><span class="dv">2</span>)</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(g)</span></code></pre></div>
<p><img src="gwas_summary_data_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>Positive selection:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>g &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span>(<span class="dt">maf=</span><span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="dv">1000</span>, <span class="fl">0.01</span>, <span class="fl">0.5</span>), <span class="dt">h2=</span><span class="fl">0.4</span>, <span class="dt">S=</span><span class="dv">2</span>)</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(g)</span></code></pre></div>
<p><img src="gwas_summary_data_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>Neutral model:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>g &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span>(<span class="dt">maf=</span><span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="dv">1000</span>, <span class="fl">0.01</span>, <span class="fl">0.5</span>), <span class="dt">h2=</span><span class="fl">0.4</span>, <span class="dt">S=</span><span class="dv">0</span>)</span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(g)</span></code></pre></div>
<p><img src="gwas_summary_data_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>Note that the genetic variance is the heritability</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(g<span class="op">$</span>beta<span class="op">^</span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>g<span class="op">$</span>maf <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span><span class="op">-</span>g<span class="op">$</span>maf))</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="co">#&gt; [1] 0.4</span></span></code></pre></div>
<p>Bringing it all together, we can simulate a set of GWAS summary stats by first specifying the true effects we want, and then obtaining sampled effects, standard errors and p-values.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># Simulate 1 million SNPs, </span></span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="co"># with 10 large effects, </span></span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="co"># a polygenic background of 10000 SNPs, </span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="co"># and all others with no effects</span></span>
<span id="cb8-5"><a href="#cb8-5"></a>param &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(</span>
<span id="cb8-6"><a href="#cb8-6"></a>    <span class="kw"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span>(<span class="dt">maf=</span><span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="dv">10</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span>), <span class="dt">h2=</span><span class="fl">0.1</span>, <span class="dt">S=</span><span class="dv">0</span>),</span>
<span id="cb8-7"><a href="#cb8-7"></a>    <span class="kw"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span>(<span class="dt">maf=</span><span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="dv">10000</span>, <span class="fl">0.01</span>, <span class="fl">0.5</span>), <span class="dt">h2=</span><span class="fl">0.3</span>, <span class="dt">S=</span><span class="dv">0</span>),</span>
<span id="cb8-8"><a href="#cb8-8"></a>    <span class="kw"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span>(<span class="dt">maf=</span><span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="dv">1000000-10-10000</span>, <span class="fl">0.01</span>, <span class="fl">0.5</span>), <span class="dt">h2=</span><span class="dv">0</span>, <span class="dt">S=</span><span class="dv">0</span>)</span>
<span id="cb8-9"><a href="#cb8-9"></a>)</span>
<span id="cb8-10"><a href="#cb8-10"></a></span>
<span id="cb8-11"><a href="#cb8-11"></a><span class="co"># Generate GWAS summary stats</span></span>
<span id="cb8-12"><a href="#cb8-12"></a>res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generate_gwas_ss.html">generate_gwas_ss</a></span>(param<span class="op">$</span>beta, param<span class="op">$</span>maf, <span class="dv">450000</span>)</span></code></pre></div>
<p>Compare to individual level data results to summary data simulations</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">100</span>)</span>
<span id="cb9-2"><a href="#cb9-2"></a>h2 &lt;-<span class="st"> </span><span class="fl">0.1</span></span>
<span id="cb9-3"><a href="#cb9-3"></a>nid &lt;-<span class="st"> </span><span class="dv">10000</span></span>
<span id="cb9-4"><a href="#cb9-4"></a>param &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span>(<span class="dt">maf=</span><span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="dv">100</span>, <span class="fl">0.01</span>, <span class="fl">0.5</span>), <span class="dt">h2=</span>h2, <span class="dt">S=</span><span class="dv">0</span>)</span>
<span id="cb9-5"><a href="#cb9-5"></a>res1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generate_gwas_ss.html">generate_gwas_ss</a></span>(param<span class="op">$</span>beta, param<span class="op">$</span>maf, nid)</span>
<span id="cb9-6"><a href="#cb9-6"></a></span>
<span id="cb9-7"><a href="#cb9-7"></a>g &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(param<span class="op">$</span>maf, <span class="cf">function</span>(x) <span class="kw"><a href="https://rdrr.io/r/stats/Binomial.html">rbinom</a></span>(nid, <span class="dv">2</span>, x)) <span class="op">%&gt;%</span></span>
<span id="cb9-8"><a href="#cb9-8"></a><span class="st">    </span>bind_cols <span class="op">%&gt;%</span></span>
<span id="cb9-9"><a href="#cb9-9"></a><span class="st">    </span>as.matrix</span>
<span id="cb9-10"><a href="#cb9-10"></a></span>
<span id="cb9-11"><a href="#cb9-11"></a>score &lt;-<span class="st"> </span>g <span class="op">%*%</span><span class="st"> </span>param<span class="op">$</span>beta</span>
<span id="cb9-12"><a href="#cb9-12"></a>err &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(nid, <span class="dt">mean=</span><span class="dv">0</span>, <span class="dt">sd=</span><span class="kw"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/cor.html">var</a></span>(score) <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span><span class="op">-</span>h2) <span class="op">/</span><span class="st"> </span>h2))</span>
<span id="cb9-13"><a href="#cb9-13"></a>y &lt;-<span class="st"> </span>score <span class="op">+</span><span class="st"> </span>err</span>
<span id="cb9-14"><a href="#cb9-14"></a>res2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gwas.html">gwas</a></span>(y, g)</span>
<span id="cb9-15"><a href="#cb9-15"></a><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(res2<span class="op">$</span>bhat <span class="op">~</span><span class="st"> </span>res1<span class="op">$</span>bhat)</span></code></pre></div>
<p><img src="gwas_summary_data_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(res2<span class="op">$</span>se, res1<span class="op">$</span>se)</span></code></pre></div>
<p><img src="gwas_summary_data_files/figure-html/unnamed-chunk-9-2.png" width="700"></p>
<p><strong>TODO: simulate effect estimates for SNPs given LD matrix</strong></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#expected-standard-error">Expected standard error</a></li>
      <li><a href="#sampling-the-effect-estimate">Sampling the effect estimate</a></li>
      <li><a href="#sampling-a-set-of-effect-estimates">Sampling a set of effect estimates</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Gibran Hemani.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
