<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simulating GWAS summary data • simulateGP</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Simulating GWAS summary data">
<meta property="og:description" content="simulateGP">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">simulateGP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/index.html">Guide</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/explodecomputer/simulateGP">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="simulate_gwas_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Simulating GWAS summary data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/explodecomputer/simulateGP/blob/master/vignettes/simulate_gwas.Rmd"><code>vignettes/simulate_gwas.Rmd</code></a></small>
      <div class="hidden name"><code>simulate_gwas.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://explodecomputer.github.io/simulateGP/">simulateGP</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></pre></div>
<p>We’d like to be able to simulate realistic GWAS summary data for a large (e.g. millions) of SNPs and representing arbitrarily large sample sizes. The simplest way to do this is to take a genetic reference dataset (e.g. UK Biobank, 500000 samples), and generate a phenotype based on the SNPs that we want to be causal. We would then run a GWAS on the simulated phenotype to obtain the GWAS summary data. However this would be very slow. The <code>simulateGP</code> package has a few methods to rapidly simulate GWAS summary data directly without going via analysis of individual level datasets.</p>
<p>Features of GWAS:</p>
<ul>
<li>Specified heritability</li>
<li>Distribution of effect sizes according to a model of natural selection</li>
<li>Effect sizes have patterns relating to linkage disequilibrium</li>
<li>LD patterns reflect specific major populations</li>
<li>Two GWAS summary datasets are correlated due to sample overlap</li>
</ul>
<div id="genetic-architecture" class="section level2">
<h2 class="hasAnchor">
<a href="#genetic-architecture" class="anchor"></a>Genetic architecture</h2>
<p>We use the following model to describe the distribution of effect sizes for each SNP <span class="math inline">\(j...M\)</span> across the genome:</p>
<p><span class="math display">\[
\beta_j \sim N(0, [2p_j (1-p_j)]^S \sigma^2_\beta)\pi + \vartheta (1 - \pi)
\]</span></p>
<p>where <span class="math inline">\(\beta_j\)</span> is the effect size, <span class="math inline">\(p_j\)</span> is the allele frequency, <span class="math inline">\(S\)</span> is the selection coefficient acting on the trait, <span class="math inline">\(\pi\)</span> is the fraction of SNPs that have an effect and <span class="math inline">\(\vartheta()\)</span> represents a null value for all SNPs that have no effect. If the phenotypic variance is 1 then <span class="math inline">\(M \sigma^2_\beta\)</span> is the heritability.</p>
<p>We can generate effect sizes for an arbitrary number of SNPs by starting with a set of allele frequencies and supplying the relevant parameters described above</p>
<p>First create a map of all the SNPs to be simulated, containing at least the SNP names and the effect allele frequencies</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="va">nsnp</span> <span class="op">&lt;-</span> <span class="fl">1000000</span>
<span class="va">map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>snp<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">nsnp</span>, af<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="va">nsnp</span>, <span class="fl">0.01</span>, <span class="fl">0.99</span><span class="op">)</span><span class="op">)</span></pre></div>
<p>Next, generate effect sizes for each SNP</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span><span class="op">(</span>map<span class="op">=</span><span class="va">map</span>, h2<span class="op">=</span><span class="fl">0.4</span>, S<span class="op">=</span><span class="op">-</span><span class="fl">0.4</span>, Pi<span class="op">=</span><span class="fl">0.005</span><span class="op">)</span></pre></div>
<p>Here there is negative selection acting on a small number of causal variants, and the overall heritability explained by all causal variants is 0.4. Observe the relationship between allele frequency and effect size for this selection coefficient:</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">params</span><span class="op">[</span><span class="va">params</span><span class="op">$</span><span class="va">beta</span><span class="op">!=</span><span class="fl">0</span>,<span class="op">]</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">af</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">beta</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<p>Finally, generate sample effect estimates, standard errors and p-values for an arbitrary sample size, e.g. for 10 million samples here</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="va">nid</span> <span class="op">&lt;-</span> <span class="fl">10000000</span>
<span class="va">ss</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_gwas_ss.html">generate_gwas_ss</a></span><span class="op">(</span><span class="va">params</span>, nid<span class="op">=</span><span class="va">nid</span><span class="op">)</span>
<span class="va">ss</span></pre></div>
<p>Note that you can generate more complex underlying genetic architectures manually. For example if there is a mixture of distributions, where some SNPs have large effects and many have small effects:</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="va">map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/rbindlist.html">rbind</a></span><span class="op">(</span>
    <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>af <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span><span class="op">)</span>, group<span class="op">=</span><span class="fl">1</span><span class="op">)</span>,
    <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>af <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">10000</span>, <span class="fl">0.01</span>, <span class="fl">0.5</span><span class="op">)</span>, group<span class="op">=</span><span class="fl">2</span><span class="op">)</span>,
    <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>af <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">1000000</span><span class="op">-</span><span class="fl">10</span><span class="op">-</span><span class="fl">10000</span>, <span class="fl">0.01</span>, <span class="fl">0.5</span><span class="op">)</span>, group<span class="op">=</span><span class="fl">3</span><span class="op">)</span>
    <span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>snp<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>

<span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/rbindlist.html">rbind</a></span><span class="op">(</span>
    <span class="fu"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span><span class="op">(</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/subset.data.table.html">subset</a></span><span class="op">(</span><span class="va">map</span>, <span class="va">group</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span>, h2<span class="op">=</span><span class="fl">0.1</span>, S<span class="op">=</span><span class="fl">0</span><span class="op">)</span>,
    <span class="fu"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span><span class="op">(</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/subset.data.table.html">subset</a></span><span class="op">(</span><span class="va">map</span>, <span class="va">group</span><span class="op">==</span><span class="fl">2</span><span class="op">)</span>, h2<span class="op">=</span><span class="fl">0.3</span>, S<span class="op">=</span><span class="fl">0</span><span class="op">)</span>,
    <span class="fu"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span><span class="op">(</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/subset.data.table.html">subset</a></span><span class="op">(</span><span class="va">map</span>, <span class="va">group</span><span class="op">==</span><span class="fl">3</span><span class="op">)</span>, h2<span class="op">=</span><span class="fl">0</span>, S<span class="op">=</span><span class="fl">0</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># Generate GWAS summary stats</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_gwas_ss.html">generate_gwas_ss</a></span><span class="op">(</span><span class="va">param</span>, <span class="fl">450000</span><span class="op">)</span>
<span class="va">res</span></pre></div>
</div>
<div id="linkage-disequilibrium" class="section level2">
<h2 class="hasAnchor">
<a href="#linkage-disequilibrium" class="anchor"></a>Linkage disequilibrium</h2>
<p>The above effects assume all SNPs are completely independent. In order for the characteristic patterns of LD around a causal variant to be generated, we need to transform the distribution of effects according to the correlation matrix amongst the SNPs.</p>
<p>We need a set of LD matrices to generate realistic LD patterns for a large set of SNPs. LD matrices are available for three major human populations for download here:</p>
<ul>
<li>Africans - <a href="http://fileserve.mrcieu.ac.uk/ld/AFR_1kg_hm3_ldobj.tar">http://fileserve.mrcieu.ac.uk/ld/AFR_1kg_hm3_ldobj.tar</a>
</li>
<li>East Asians - <a href="http://fileserve.mrcieu.ac.uk/ld/EAS_1kg_hm3_ldobj.tar">http://fileserve.mrcieu.ac.uk/ld/EAS_1kg_hm3_ldobj.tar</a>
</li>
<li>Europeans - <a href="http://fileserve.mrcieu.ac.uk/ld/EUR_1kg_hm3_ldobj.tar">http://fileserve.mrcieu.ac.uk/ld/EUR_1kg_hm3_ldobj.tar</a>
</li>
</ul>
<p>The above are large (~5Gb) <code>.tar</code> files. Download one and extract, it should contain a directory called <code>&lt;POP&gt;_1kg_hm3/</code>, which contains a <code>map.rds</code> and a number of <code>ldobj_&lt;region&gt;.rds</code> files. We will use the <code>map.rds</code> as our genome scaffold - it contains SNPs, chromosomes, positions, alleles and allele frequencies for the specific population. We can then select causal variants for the trait as above, and then introduce the LD transformation before sampling the effect estimates / SEs / p-values etc from the LD-aware effect sizes.</p>
<p>Set the LD matrix data directory:</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="va">datadir</span> <span class="op">&lt;-</span> <span class="st">"~/data/ld_files/ldmat/AFR_1kg_hm3"</span>
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">datadir</span><span class="op">)</span></pre></div>
<p>Begin with a small number of SNPs for illustration. E.g. let’s use the first region in the set of regions downloaded</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="va">fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">datadir</span>, full.names<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"ldobj_chr"</span>, <span class="va">.</span>, value<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="op">{</span><span class="va">.</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">}</span>
<span class="va">ldobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="va">fn</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">ldobj</span><span class="op">)</span></pre></div>
<p>We see it contains two items in a list, the <code>map</code> and the <code>ld</code> objects. The <code>map</code> is a data frame containing information about the SNPs in the region. The <code>ld</code> is a matrix containing LD correlations amongst all the SNPs in the region. To use this to transform our data - let’s simulate a single causal variant in the region</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="va">params</span> <span class="op">&lt;-</span> <span class="va">ldobj</span><span class="op">$</span><span class="va">map</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span><span class="op">(</span>h2<span class="op">=</span><span class="fl">0.01</span>, S<span class="op">=</span><span class="fl">0</span>, Pi<span class="op">=</span><span class="fl">1</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span></pre></div>
<p>Now add LD to the effects</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="va">params_ld</span> <span class="op">&lt;-</span> <span class="va">params</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/add_ld_to_params.html">add_ld_to_params</a></span><span class="op">(</span>ldobj<span class="op">=</span><span class="va">ldobj</span><span class="op">)</span></pre></div>
<p>We can now sample effect estimates for either the the raw effects or the LD-aware effects</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="va">ss_raw</span> <span class="op">&lt;-</span> <span class="va">params</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/generate_gwas_ss.html">generate_gwas_ss</a></span><span class="op">(</span>nid<span class="op">=</span><span class="fl">10000</span><span class="op">)</span>
<span class="va">ss_ld</span> <span class="op">&lt;-</span> <span class="va">params_ld</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/generate_gwas_ss.html">generate_gwas_ss</a></span><span class="op">(</span>nid<span class="op">=</span><span class="fl">10000</span><span class="op">)</span></pre></div>
<p>Compare the associations in the region. With no LD:</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ss_raw</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">pos</span>, y<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<p>Now when LD is added:</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ss_ld</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">pos</span>, y<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<p>The same process can be achieved for more complex architectures across the entire genome. It takes about 2 minutes to generate the LD-aware effect sizes, from which multiple independent sample effect estimates can be drawn nearly instantaneously.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="va">map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"map.rds"</span><span class="op">)</span><span class="op">)</span>
<span class="va">params</span> <span class="op">&lt;-</span> <span class="va">map</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span><span class="op">(</span>h2<span class="op">=</span><span class="fl">0.5</span>, S<span class="op">=</span><span class="op">-</span><span class="fl">0.3</span>, Pi<span class="op">=</span><span class="fl">1000</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/add_ld_to_params.html">add_ld_to_params</a></span><span class="op">(</span>ldobjdir<span class="op">=</span><span class="va">datadir</span><span class="op">)</span>

<span class="va">ss</span> <span class="op">&lt;-</span> <span class="va">params</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/generate_gwas_ss.html">generate_gwas_ss</a></span><span class="op">(</span>nid<span class="op">=</span><span class="fl">1000000</span><span class="op">)</span></pre></div>
</div>
<div id="sample-overlap" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-overlap" class="anchor"></a>Sample overlap</h2>
<p>Finally, we can draw joint estimates of two different GWASs that share some amount of sample overlap using the <code>summary_set</code> function. The steps are:</p>
<p>For a set of variants, generate the true effect sizes for each of two different datasets. Here we make a simple example using the single small region as above. Trait 1 has a causal effect of 0.3 on trait 2.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="va">params1</span> <span class="op">&lt;-</span> <span class="va">ldobj</span><span class="op">$</span><span class="va">map</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span><span class="op">(</span>h2<span class="op">=</span><span class="fl">0.01</span>, S<span class="op">=</span><span class="fl">0</span>, Pi<span class="op">=</span><span class="fl">3</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span>

<span class="va">params2</span> <span class="op">&lt;-</span> <span class="va">params2</span>
<span class="va">params2</span><span class="op">$</span><span class="va">beta</span> <span class="op">&lt;-</span> <span class="va">params1</span><span class="op">$</span><span class="va">beta</span> <span class="op">*</span> <span class="fl">0.3</span></pre></div>
<p>Now sample the effect estimates jointly. We specify the sample sizes of each GWAS, and the number of overlapping samples. We also specify an observational correlation, which represents the total association due to the causal relationship and any confounding.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="va">ss</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summary_set.html">summary_set</a></span><span class="op">(</span>
    beta_gx<span class="op">=</span><span class="va">params1</span><span class="op">$</span><span class="va">beta</span>,
    beta_gy<span class="op">=</span><span class="va">params2</span><span class="op">$</span><span class="va">beta</span>,
    af<span class="op">=</span><span class="va">params1</span><span class="op">$</span><span class="va">af</span>,
    n_gx<span class="op">=</span><span class="fl">100000</span>,
    n_gy<span class="op">=</span><span class="fl">10000</span>,
    n_overlap<span class="op">=</span><span class="fl">5000</span>,
    cor_xy<span class="op">=</span><span class="fl">0.5</span>
<span class="op">)</span></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Gibran Hemani, John Ferguson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
