<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simple two-sample Mendelian randomisation examples • simulateGP</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Simple two-sample Mendelian randomisation examples">
<meta property="og:description" content="simulateGP">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">simulateGP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/index.html">Guide</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/explodecomputer/simulateGP">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="twosamplemr_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Simple two-sample Mendelian randomisation examples</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/explodecomputer/simulateGP/blob/master/vignettes/twosamplemr.Rmd"><code>vignettes/twosamplemr.Rmd</code></a></small>
      <div class="hidden name"><code>twosamplemr.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://explodecomputer.github.io/simulateGP/">simulateGP</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TwoSampleMR</span><span class="op">)</span>
<span class="co">#&gt; TwoSampleMR version 0.5.6 </span>
<span class="co">#&gt; [&gt;] New: Option to use non-European LD reference panels for clumping etc</span>
<span class="co">#&gt; [&gt;] Some studies temporarily quarantined to verify effect allele</span>
<span class="co">#&gt; [&gt;] See news(package='TwoSampleMR') and https://gwas.mrcieu.ac.uk for further details</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'TwoSampleMR'</span>
<span class="co">#&gt; The following objects are masked from 'package:simulateGP':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     allele_frequency, contingency, get_population_allele_frequency</span></pre></div>
<p>A convenient approach to performing Mendelian randomisation is to use GWAS summary data, often known as summary MR or 2-sample MR. This figure outlines the approach:</p>
<div class="figure">
<img src="2smr.png" alt=""><p class="caption">2smr</p>
</div>
<p>In this vignette we simulate some examples of how to simulate the underlying GWAS summary data for 2-sample MR.</p>
<div id="generating-summary-data-from-individual-level-data" class="section level2">
<h2 class="hasAnchor">
<a href="#generating-summary-data-from-individual-level-data" class="anchor"></a>Generating summary data from individual level data</h2>
<p>Here we want to achieve the following:</p>
<ol style="list-style-type: decimal">
<li>Simulate some genetic or confounding variables</li>
<li>Simulate exposures that are influenced by (1)</li>
<li>Simulate the outcomes that are influenced by (1) and (2)</li>
<li>Get the SNP-exposure and SNP-outcome effects and standard errors for these simulations, for use in further analyses e.g. using the <a href="https://github.com/MRCIEU/TwoSampleMR">TwoSampleMR</a> package</li>
</ol>
<p>Here is how to do 1-3:</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co"># Genotypes for 20 SNPs and 10000 individuals, where the SNPs have MAF = 0.5:</span>
<span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_geno.html">make_geno</a></span><span class="op">(</span><span class="fl">10000</span>, <span class="fl">20</span>, <span class="fl">0.5</span><span class="op">)</span>

<span class="co"># These SNPs instrument some exposure, and together explain 5% of the variance</span>
<span class="va">effs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/choose_effects.html">choose_effects</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">0.05</span><span class="op">)</span>

<span class="co"># Create X</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_phen.html">make_phen</a></span><span class="op">(</span><span class="va">effs</span>, <span class="va">g</span><span class="op">)</span>

<span class="co"># Check that the SNPs explain 5% of the variance in x</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ieugwasr/man/cor.html">cor</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">g</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.06101182</span>

<span class="co"># Create Y, which is negatively influenced by x, explaining 10% of the variance in Y#</span>
<span class="co"># so if the variances of x and y are both 1 then the effect size is about -0.31</span>
<span class="va">beta_xy</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="fl">0.1</span><span class="op">)</span>
<span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_phen.html">make_phen</a></span><span class="op">(</span><span class="va">beta_xy</span>, <span class="va">x</span><span class="op">)</span></pre></div>
<p>We now have an X and Y, and the genotypes. To perform 2-sample MR on this we need to obtain the summary data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_effs.html">get_effs</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">g</span><span class="op">)</span></pre></div>
<p>Perform MR</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="fu">TwoSampleMR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/mr.html">mr</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span>
<span class="co">#&gt; Analysing 'X' on 'Y'</span>
<span class="co">#&gt;   id.exposure id.outcome outcome exposure                    method nsnp</span>
<span class="co">#&gt; 1           X          Y       Y        X                  MR Egger   20</span>
<span class="co">#&gt; 2           X          Y       Y        X           Weighted median   20</span>
<span class="co">#&gt; 3           X          Y       Y        X Inverse variance weighted   20</span>
<span class="co">#&gt; 4           X          Y       Y        X               Simple mode   20</span>
<span class="co">#&gt; 5           X          Y       Y        X             Weighted mode   20</span>
<span class="co">#&gt;            b         se         pval</span>
<span class="co">#&gt; 1 -0.3067359 0.06945463 3.331980e-04</span>
<span class="co">#&gt; 2 -0.3319027 0.06298175 1.365552e-07</span>
<span class="co">#&gt; 3 -0.3440340 0.04093057 4.267401e-17</span>
<span class="co">#&gt; 4 -0.4273474 0.10099633 4.518605e-04</span>
<span class="co">#&gt; 5 -0.3732034 0.07933612 1.544933e-04</span></pre></div>
<p>The effect size is approx -0.31 as we simulated.</p>
<div id="more-complex-example" class="section level3">
<h3 class="hasAnchor">
<a href="#more-complex-example" class="anchor"></a>More complex example</h3>
<p>Let’s simulate an instance where there are two traits that have different effects on the outcome, but they share some of the same instruments</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co"># Simulate 100 genotypes</span>
<span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_geno.html">make_geno</a></span><span class="op">(</span><span class="fl">100000</span>, <span class="fl">80</span>, <span class="fl">0.5</span><span class="op">)</span>

<span class="co"># Choose effect sizes for instruments for each trait</span>
<span class="va">effs1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/choose_effects.html">choose_effects</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">0.05</span><span class="op">)</span>
<span class="va">effs2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/choose_effects.html">choose_effects</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">0.05</span><span class="op">)</span>

<span class="co"># Create X1 and X2, where they overlap some variants</span>
<span class="va">x1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_phen.html">make_phen</a></span><span class="op">(</span><span class="va">effs1</span>, <span class="va">g</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span><span class="op">)</span>
<span class="va">x2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_phen.html">make_phen</a></span><span class="op">(</span><span class="va">effs2</span>, <span class="va">g</span><span class="op">[</span>,<span class="fl">31</span><span class="op">:</span><span class="fl">80</span><span class="op">]</span><span class="op">)</span>

<span class="co"># Create Y - x1 has a -0.3 influence on it and x2 has a +0.3 influence on it</span>
<span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_phen.html">make_phen</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.3</span>, <span class="fl">0.3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">x1</span>, <span class="va">x2</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Perform separate MR on each</span>
<span class="va">dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_effs.html">get_effs</a></span><span class="op">(</span><span class="va">x1</span>, <span class="va">y</span>, <span class="va">g</span><span class="op">)</span>
<span class="va">dat2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_effs.html">get_effs</a></span><span class="op">(</span><span class="va">x2</span>, <span class="va">y</span>, <span class="va">g</span><span class="op">)</span>

<span class="fu">TwoSampleMR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/mr.html">mr</a></span><span class="op">(</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/subset.data.table.html">subset</a></span><span class="op">(</span><span class="va">dat1</span>, <span class="va">pval.exposure</span> <span class="op">&lt;</span> <span class="fl">5e-8</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Analysing 'X' on 'Y'</span>
<span class="co">#&gt;   id.exposure id.outcome outcome exposure                    method nsnp</span>
<span class="co">#&gt; 1           X          Y       Y        X                  MR Egger   28</span>
<span class="co">#&gt; 2           X          Y       Y        X           Weighted median   28</span>
<span class="co">#&gt; 3           X          Y       Y        X Inverse variance weighted   28</span>
<span class="co">#&gt; 4           X          Y       Y        X               Simple mode   28</span>
<span class="co">#&gt; 5           X          Y       Y        X             Weighted mode   28</span>
<span class="co">#&gt;            b         se         pval</span>
<span class="co">#&gt; 1 -0.2956305 0.08348686 1.528040e-03</span>
<span class="co">#&gt; 2 -0.3241051 0.02284613 1.111657e-45</span>
<span class="co">#&gt; 3 -0.3382881 0.03058010 1.910047e-28</span>
<span class="co">#&gt; 4 -0.3200901 0.03695778 2.820664e-09</span>
<span class="co">#&gt; 5 -0.3172545 0.03543751 1.442211e-09</span>
<span class="fu">TwoSampleMR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/mr.html">mr</a></span><span class="op">(</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/subset.data.table.html">subset</a></span><span class="op">(</span><span class="va">dat2</span>, <span class="va">pval.exposure</span> <span class="op">&lt;</span> <span class="fl">5e-8</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Analysing 'X' on 'Y'</span>
<span class="co">#&gt;   id.exposure id.outcome outcome exposure                    method nsnp</span>
<span class="co">#&gt; 1           X          Y       Y        X                  MR Egger   29</span>
<span class="co">#&gt; 2           X          Y       Y        X           Weighted median   29</span>
<span class="co">#&gt; 3           X          Y       Y        X Inverse variance weighted   29</span>
<span class="co">#&gt; 4           X          Y       Y        X               Simple mode   29</span>
<span class="co">#&gt; 5           X          Y       Y        X             Weighted mode   29</span>
<span class="co">#&gt;           b         se         pval</span>
<span class="co">#&gt; 1 0.2611408 0.09359298 9.546175e-03</span>
<span class="co">#&gt; 2 0.2874271 0.02700271 1.852153e-26</span>
<span class="co">#&gt; 3 0.2912836 0.04023173 4.482034e-13</span>
<span class="co">#&gt; 4 0.2648308 0.04439272 2.007401e-06</span>
<span class="co">#&gt; 5 0.3094855 0.02887952 2.044689e-11</span>

<span class="co"># Do multivariable MR</span>
<span class="co"># First get the effects for x1, x2 and y, and put them in mv format</span>
<span class="va">mvdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_mvdat.html">make_mvdat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">x1</span>, <span class="va">x2</span><span class="op">)</span>, <span class="va">y</span>, <span class="va">g</span><span class="op">)</span>
<span class="co">#&gt; There are 2 exposures</span>
<span class="co">#&gt; Harmonising x1 (x1) and y (y)</span>

<span class="co"># Perform MV MR</span>
<span class="fu">TwoSampleMR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/mv_multiple.html">mv_multiple</a></span><span class="op">(</span><span class="va">mvdat</span><span class="op">)</span>
<span class="co">#&gt; $result</span>
<span class="co">#&gt;   id.exposure exposure id.outcome outcome nsnp          b         se</span>
<span class="co">#&gt; 1          x1       x1          y       y   28 -0.3369279 0.01212065</span>
<span class="co">#&gt; 2          x2       x2          y       y   29  0.2976037 0.01228111</span>
<span class="co">#&gt;            pval</span>
<span class="co">#&gt; 1 4.606670e-170</span>
<span class="co">#&gt; 2 1.008165e-129</span></pre></div>
</div>
</div>
<div id="generating-summary-data-directly" class="section level2">
<h2 class="hasAnchor">
<a href="#generating-summary-data-directly" class="anchor"></a>Generating summary data directly</h2>
<p>For large scale analyses it is more convenient to just generate the summary data directly. For this we need to specify the genotype effects on <span class="math inline">\(x\)</span>, determine the standard errors expected based on the sample size, and then determine the genotype effect on <span class="math inline">\(y\)</span> according to causal effects of <span class="math inline">\(x\)</span> and pleiotropic paths, and determine the sample sizes for <span class="math inline">\(y\)</span> also. We can additionally specify the extent of sample overlap.</p>
<div id="no-sample-overlap-case" class="section level3">
<h3 class="hasAnchor">
<a href="#no-sample-overlap-case" class="anchor"></a>No sample overlap case</h3>
<ol style="list-style-type: decimal">
<li>Choose true effects of instruments on <span class="math inline">\(x\)</span>
</li>
<li>Generate summary data for <span class="math inline">\(x\)</span>
</li>
<li>Determine effects of instruments on <span class="math inline">\(y\)</span>
</li>
<li>Generate summary data for <span class="math inline">\(y\)</span>
</li>
<li>Perform two-sample MR</li>
</ol>
<p>Choose some number of SNPs and the effect allele frequencies and the variance explained in <span class="math inline">\(x\)</span> by the SNPs</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="va">nsnp</span> <span class="op">&lt;-</span> <span class="fl">100</span>
<span class="va">af</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>
<span class="va">rsq_gx</span> <span class="op">&lt;-</span> <span class="fl">0.4</span></pre></div>
<p>Generate some effects that correspond to these values</p>
<div class="sourceCode" id="cb7"><pre class="downlit">

<span class="va">b_gx</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>af<span class="op">=</span><span class="va">af</span>, snp<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">nsnp</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span><span class="op">(</span>h2<span class="op">=</span><span class="va">rsq_gx</span><span class="op">)</span></pre></div>
<p>Generate sample effects for 1 million samples</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="va">nid_x</span> <span class="op">&lt;-</span> <span class="fl">1000000</span>
<span class="va">bhat_gx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_gwas_ss.html">generate_gwas_ss</a></span><span class="op">(</span><span class="va">b_gx</span>, nid<span class="op">=</span><span class="va">nid_x</span><span class="op">)</span></pre></div>
<p>Assume <span class="math inline">\(x\)</span> has an effect on <span class="math inline">\(y\)</span> of -0.3</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="va">b_xy</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fl">0.3</span></pre></div>
<p>Simulate some balanced pleiotropy - assume that 50% of the instruments have a direct effect on <span class="math inline">\(y\)</span> and they explain 2% of the variance through this path</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="va">plei_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">nsnp</span>, <span class="va">nsnp</span><span class="op">*</span><span class="fl">0.5</span><span class="op">)</span>
<span class="va">b_gy_plei</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>af<span class="op">=</span><span class="va">af</span><span class="op">[</span><span class="va">plei_index</span><span class="op">]</span>, snp<span class="op">=</span><span class="va">plei_index</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span><span class="op">(</span>h2<span class="op">=</span><span class="fl">0.02</span><span class="op">)</span></pre></div>
<p>Sum the genetic effects on <span class="math inline">\(y\)</span></p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="va">b_gy</span> <span class="op">&lt;-</span> <span class="va">b_gx</span>
<span class="va">b_gy</span><span class="op">$</span><span class="va">beta</span> <span class="op">&lt;-</span> <span class="va">b_gy</span><span class="op">$</span><span class="va">beta</span> <span class="op">*</span> <span class="va">b_xy</span>
<span class="va">b_gy</span><span class="op">$</span><span class="va">beta</span><span class="op">[</span><span class="va">plei_index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">b_gy</span><span class="op">$</span><span class="va">beta</span><span class="op">[</span><span class="va">plei_index</span><span class="op">]</span> <span class="op">+</span> <span class="va">b_gy_plei</span><span class="op">$</span><span class="va">beta</span></pre></div>
<p>Generate summary data for g-y associations, assuming sample size of 800000</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="va">nid_y</span> <span class="op">&lt;-</span> <span class="fl">800000</span>
<span class="va">bhat_gy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_gwas_ss.html">generate_gwas_ss</a></span><span class="op">(</span><span class="va">b_gy</span>, nid<span class="op">=</span><span class="va">nid_y</span><span class="op">)</span></pre></div>
<p>Perform MR</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_exp_out.html">merge_exp_out</a></span><span class="op">(</span><span class="va">bhat_gx</span>, <span class="va">bhat_gy</span><span class="op">)</span></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="fu">TwoSampleMR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/mr.html">mr</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span>
<span class="co">#&gt; Analysing 'X' on 'Y'</span>
<span class="co">#&gt;   id.exposure id.outcome outcome exposure                    method nsnp</span>
<span class="co">#&gt; 1           X          Y       Y        X                  MR Egger  100</span>
<span class="co">#&gt; 2           X          Y       Y        X           Weighted median  100</span>
<span class="co">#&gt; 3           X          Y       Y        X Inverse variance weighted  100</span>
<span class="co">#&gt; 4           X          Y       Y        X               Simple mode  100</span>
<span class="co">#&gt; 5           X          Y       Y        X             Weighted mode  100</span>
<span class="co">#&gt;            b          se         pval</span>
<span class="co">#&gt; 1 -0.3002642 0.035778493 3.665339e-13</span>
<span class="co">#&gt; 2 -0.2996721 0.003766891 0.000000e+00</span>
<span class="co">#&gt; 3 -0.2660478 0.022431174 1.895495e-32</span>
<span class="co">#&gt; 4 -0.3087840 0.011766201 2.214902e-46</span>
<span class="co">#&gt; 5 -0.3087840 0.012601099 8.003348e-44</span></pre></div>
</div>
<div id="sample-overlap-case" class="section level3">
<h3 class="hasAnchor">
<a href="#sample-overlap-case" class="anchor"></a>Sample overlap case</h3>
<p>It is of course possible to simulate varying degrees of sample overlap by simulating two individual-level datasets with the same underlying DAG, and then selecting which individuals contribute to the exposure and outcome effects e.g.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="va">nid</span> <span class="op">&lt;-</span> <span class="fl">10000</span>
<span class="va">nsnp</span> <span class="op">&lt;-</span> <span class="fl">100</span>
<span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_geno.html">make_geno</a></span><span class="op">(</span><span class="va">nid</span>, <span class="va">nsnp</span>, <span class="fl">0.5</span><span class="op">)</span>
<span class="va">u</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">nid</span><span class="op">)</span>
<span class="va">beta_gx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/choose_effects.html">choose_effects</a></span><span class="op">(</span><span class="va">nsnp</span>, <span class="fl">0.05</span><span class="op">)</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_phen.html">make_phen</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">beta_gx</span>, <span class="fl">0.1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">g</span>, <span class="va">u</span><span class="op">)</span><span class="op">)</span>
<span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_phen.html">make_phen</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">u</span><span class="op">)</span><span class="op">)</span></pre></div>
<p>Complete sample overlap</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_effs.html">get_effs</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">g</span><span class="op">)</span></pre></div>
<p>No sample overlap</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="va">id1</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="va">nid</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span>
<span class="va">id2</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">nid</span><span class="op">/</span><span class="fl">2</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="va">nid</span>
<span class="va">exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gwas.html">gwas</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="va">id1</span><span class="op">]</span>, <span class="va">g</span><span class="op">[</span><span class="va">id1</span>,<span class="op">]</span><span class="op">)</span>
<span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gwas.html">gwas</a></span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="va">id2</span><span class="op">]</span>, <span class="va">g</span><span class="op">[</span><span class="va">id2</span>,<span class="op">]</span><span class="op">)</span>
<span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_exp_out.html">merge_exp_out</a></span><span class="op">(</span><span class="va">exp</span>, <span class="va">out</span><span class="op">)</span></pre></div>
<p>Partial sample overlap</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="va">id1</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">nid</span><span class="op">/</span><span class="fl">1.5</span><span class="op">)</span>
<span class="va">id2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">nid</span><span class="op">/</span><span class="fl">1.5</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="va">nid</span>
<span class="va">exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gwas.html">gwas</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="va">id1</span><span class="op">]</span>, <span class="va">g</span><span class="op">[</span><span class="va">id1</span>,<span class="op">]</span><span class="op">)</span>
<span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gwas.html">gwas</a></span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="va">id2</span><span class="op">]</span>, <span class="va">g</span><span class="op">[</span><span class="va">id2</span>,<span class="op">]</span><span class="op">)</span>
<span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_exp_out.html">merge_exp_out</a></span><span class="op">(</span><span class="va">exp</span>, <span class="va">out</span><span class="op">)</span></pre></div>
<hr>
<p>An approach to doing this directly, bypassing simulation of individuals, is also implemented in the <code>summary_set</code> function e.g.</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="va">nsnp</span> <span class="op">&lt;-</span> <span class="fl">100</span>
<span class="va">af</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="va">nsnp</span>, <span class="fl">0.01</span>, <span class="fl">0.99</span><span class="op">)</span>
<span class="va">beta_gx</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>af<span class="op">=</span><span class="va">af</span>, snp<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">nsnp</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/generate_gwas_params.html">generate_gwas_params</a></span><span class="op">(</span>h2<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span>
<span class="va">beta_gy</span> <span class="op">&lt;-</span> <span class="va">beta_gx</span>
<span class="va">beta_gy</span><span class="op">$</span><span class="va">beta</span> <span class="op">&lt;-</span> <span class="va">beta_gy</span><span class="op">$</span><span class="va">beta</span> <span class="op">*</span> <span class="fl">0.3</span>

<span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summary_set.html">summary_set</a></span><span class="op">(</span>
    beta_gx <span class="op">=</span> <span class="va">beta_gx</span><span class="op">$</span><span class="va">beta</span>,
    beta_gy <span class="op">=</span> <span class="va">beta_gy</span><span class="op">$</span><span class="va">beta</span>,
    af <span class="op">=</span> <span class="va">beta_gx</span><span class="op">$</span><span class="va">af</span>,
    n_gx <span class="op">=</span> <span class="fl">10000</span>,
    n_gy <span class="op">=</span> <span class="fl">10000</span>,
    n_overlap <span class="op">=</span> <span class="fl">10000</span>,
    cor_xy <span class="op">=</span> <span class="fl">0.6</span>,
    prev_y <span class="op">=</span> <span class="cn">NA</span>,
    sigma_x <span class="op">=</span> <span class="fl">1</span>,
    sigma_y <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span>

<span class="va">dat</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">pval.exposure</span> <span class="op">&lt;</span> <span class="fl">5e-8</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">TwoSampleMR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/mr.html">mr</a></span><span class="op">(</span><span class="va">.</span>, method_list<span class="op">=</span><span class="st">"mr_ivw"</span><span class="op">)</span>
<span class="co">#&gt; Analysing 'X' on 'Y'</span>
<span class="co">#&gt;   id.exposure id.outcome outcome exposure                    method nsnp</span>
<span class="co">#&gt; 1           X          Y       Y        X Inverse variance weighted   39</span>
<span class="co">#&gt;           b         se          pval</span>
<span class="co">#&gt; 1 0.3186653 0.01477096 3.168257e-103</span></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Gibran Hemani, John Ferguson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
