<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ simulateGP</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">simulateGP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/fast_linreg.html">Performing fast linear regression</a>
    </li>
    <li>
      <a href="../articles/gwas_summary_data.html">Simulating GWAS summary data</a>
    </li>
    <li>
      <a href="../articles/individual_gwas.html">Simulating individual level GWAS data</a>
    </li>
    <li>
      <a href="../articles/risk.html">Translating between observed and liability models</a>
    </li>
    <li>
      <a href="../articles/simulate_individual_2smr.html">Simulate two-sample MR from individual level data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="add_u.html">add_u()</a></code> </p>
        </td>
        <td><p>Add confounder variables and their instruments</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="allele_frequency.html">allele_frequency()</a></code> </p>
        </td>
        <td><p>Estimate allele frequency from SNP</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ascertain_samples.html">ascertain_samples()</a></code> </p>
        </td>
        <td><p>Ascertain some proportion of cases and controls from binary phenotype</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="choose_effects.html">choose_effects()</a></code> </p>
        </td>
        <td><p>Get vector of effects that explain some amount of variance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="contingency.html">contingency()</a></code> </p>
        </td>
        <td><p>Obtain 2x2 contingency table from marginal parameters and odds ratio</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="create_system.html">create_system()</a></code> </p>
        </td>
        <td><p>Wrapper for simulation pipeline</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="estimate_system_effects.html">estimate_system_effects()</a></code> </p>
        </td>
        <td><p>Estimate the effects of all SNPs on all phenotypes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="expected_gwas.html">expected_gwas()</a></code> </p>
        </td>
        <td><p>For known effect size, sample size and variances, what is the expected association?</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="expected_mse.html">expected_mse()</a></code> </p>
        </td>
        <td><p>Calculate expected MSE</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="expected_se.html">expected_se()</a></code> </p>
        </td>
        <td><p>Expected se given beta, maf, n and vy</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="expected_ssx.html">expected_ssx()</a></code> </p>
        </td>
        <td><p>Calculate expected SSX</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fast_assoc.html">fast_assoc()</a></code> </p>
        </td>
        <td><p>Get summary statistics in simple linear regression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_effs.html">get_effs()</a></code> </p>
        </td>
        <td><p>Get effs for two traits and make dat format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_population_allele_frequency.html">get_population_allele_frequency()</a></code> </p>
        </td>
        <td><p>Estimate the allele frequency in population from case/control summary data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gwas.html">gwas()</a></code> </p>
        </td>
        <td><p>Perform association of many SNPs against phenotype</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gwas_se.html">gwas_se()</a></code> </p>
        </td>
        <td><p>Get the expected se for a gwas given n, h2, beta maf</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gx_to_gp.html">gx_to_gp()</a></code> </p>
        </td>
        <td><p>Translate risk from liability to probability scale</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="init_parameters.html">init_parameters()</a></code> </p>
        </td>
        <td><p>Choose initial parameters for direct effects on X and Y</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lor_to_rsq.html">lor_to_rsq()</a></code> </p>
        </td>
        <td><p>Estimate proportion of variance of liability explained by SNP in general population</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_dat.html">make_dat()</a></code> </p>
        </td>
        <td><p>Organise outputs from <code>gwas</code> into harmonised dat format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_geno.html">make_geno()</a></code> </p>
        </td>
        <td><p>Create genotype matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_mvdat.html">make_mvdat()</a></code> </p>
        </td>
        <td><p>Take several exposures and one outcome and make the data required for multivariable MR</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_phen.html">make_phen()</a></code> </p>
        </td>
        <td><p>Simulate variable based on the influences of other variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="range01.html">range01()</a></code> </p>
        </td>
        <td><p>Scale variable to have range between 0 and 1</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="recode_dat.html">recode_dat()</a></code> </p>
        </td>
        <td><p>Recode data to make every effect on x positive</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="recode_dat_intercept.html">recode_dat_intercept()</a></code> </p>
        </td>
        <td><p>Intercept recoding to have every effect on x positive</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="recode_dat_simple.html">recode_dat_simple()</a></code> </p>
        </td>
        <td><p>Simple recoding to have every effect on x positive</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="risk_cross_plot.html">risk_cross_plot()</a></code> </p>
        </td>
        <td><p>Plot liability vs probability disease risk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="risk_simulation.html">risk_simulation()</a></code> </p>
        </td>
        <td><p>Make simulation to compare disease and liability scales</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sample_beta.html">sample_beta()</a></code> </p>
        </td>
        <td><p>Sample beta values given standard errors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sample_system_effects.html">sample_system_effects()</a></code> </p>
        </td>
        <td><p>Sample the actual effects based on initial parameters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="simulate_population.html">simulate_population()</a></code> </p>
        </td>
        <td><p>Simulate individual level data from initial parameters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_system.html">test_system()</a></code> </p>
        </td>
        <td><p>Apply MR tests to system</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="theoretical_gwas.html">theoretical_gwas()</a></code> </p>
        </td>
        <td><p>Create a theoretical GWAS dataset</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="y_to_binary.html">y_to_binary()</a></code> </p>
        </td>
        <td><p>Convert continuous trait to binary</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Gibran Hemani.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


